internals/overlord/planstate/manager.go

```diff
@@ -0,0 +1,238 @@
+	return m.plan, nil
+}
+
+// AppendLayer takes a Layer, appends it to the plan's layers and updates the
+// layer.Order field to the new order. If a layer with layer.Label already
+// exists, return an error of type *LabelExists.
+func (m *PlanManager) AppendLayer(layer *plan.Layer) error {
+	m.planLock.Lock()
+	defer m.planLock.Unlock()
+
+	index, _ := findLayer(m.plan.Layers, layer.Label)
+	if index >= 0 {
+		return &LabelExists{Label: layer.Label}
+	}
+
+	return m.appendLayer(layer)
+}
+
+// CombineLayer takes a Layer, combines it to an existing layer that has the
+// same label. If no existing layer has the label, append a new one. In either
+// case, update the layer.Order field to the new order.
+func (m *PlanManager) CombineLayer(layer *plan.Layer) error {
+	m.planLock.Lock()
+	defer m.planLock.Unlock()
+
+	index, found := findLayer(m.plan.Layers, layer.Label)
+	if index < 0 {
+		// No layer found with this label, append new one.
+		return m.appendLayer(layer)
+	}
+
+	// Layer found with this label, combine into that one.
+	combined, err := plan.CombineLayers(found, layer)
+	if err != nil {
+		return err
+	}
+	combined.Order = found.Order
+	combined.Label = found.Label
+
+	// Insert combined layer back into plan's layers list.
+	newLayers := make([]*plan.Layer, len(m.plan.Layers))
+	copy(newLayers, m.plan.Layers)
+	newLayers[index] = combined
+	err = m.updatePlanLayers(newLayers)
+	if err != nil {
+		return err
+	}
+	layer.Order = found.Order
+	return nil
+}
+
+func (m *PlanManager) appendLayer(layer *plan.Layer) error {
+	newOrder := 1
+	if len(m.plan.Layers) > 0 {
+		last := m.plan.Layers[len(m.plan.Layers)-1]
+		newOrder = last.Order + 1
+	}
+
+	newLayers := append(m.plan.Layers, layer)
+	err := m.updatePlanLayers(newLayers)
+	if err != nil {
+		return err
+	}
+	layer.Order = newOrder
+	return nil
+}
+
+func (m *PlanManager) updatePlanLayers(layers []*plan.Layer) error {
+	combined, err := plan.CombineLayers(layers...)
+	if err != nil {
+		return err
+	}
+	p := &plan.Plan{
+		Layers:     layers,
+		Services:   combined.Services,
+		Checks:     combined.Checks,
+		LogTargets: combined.LogTargets,
+	}
+	p.Validate()
+	if err != nil {
+		return err
+	}
+	m.planChanged(p)
+	return nil
+}
+
+// findLayer returns the index (in layers) of the layer with the given label,
+// or returns -1, nil if there's no layer with that label.
+func findLayer(layers []*plan.Layer, label string) (int, *plan.Layer) {
+	for i, layer := range layers {
+		if layer.Label == label {
+			return i, layer
+		}
+	}
+	return -1, nil
+}
+
+// Ensure implements StateManager.Ensure.
+func (m *PlanManager) Ensure() error {
+	return nil
+}
+
+// SetServiceArgs sets the service arguments provided by "pebble run --args"
+// to their respective services. It adds a new layer in the plan, the layer
+// consisting of services with commands having their arguments changed.
+//
+// TODO: This functionality must be redesigned (moved out of the plan manager)
```

Comment:
I've been trying to reserve "TODO" for things that really need to be done soon or bug fixes (though I realise that's not always how it's used). This is sligtly non-optimal, but I'd rather say:
```suggestion
// NOTE: This functionality should be redesigned (moved out of the plan manager)
```

Reply:
@flotter, do you already have some ideas about where this should go?

Reply:
Yes, I absolutely do. But I removed the proposal from this PR since I think it will require some design related debate and ideas from Ben and potentially Gustavo, so will be a follow-on PR. But it will happen.

Reply:
Improved the wording as per Ben's suggestion. Resolving for now. 

@rebornplusplus This topic will naturally come up again because I am working on another related PR that will allow external managers to add plan sections, which means the Plan Manager cannot have schema knowledge. So I will resolve this comment now, but the conversation will not be lost.